### PR TITLE
[Relay][QNN] Add unit test for int8

### PR DESCRIPTION
**Pull request by zhiics**
This PR fixes the bug in casting from `int` to `uint` in llvm codegen where `to` should be used as the destination type instead of `from`.  It also fixed the prefix of unit test in test_op_qnn_conv2d.py.

@anijain2305 @icemelon9

### CODE REVIEW

**Comment by anijain2305** [comment]
Thanks @zhiics This was blocking QNN - https://discuss.tvm.ai/t/segfault-in-llvm/3567
:)

**Comment by zhiics** [comment]
@anijain2305 hmm, sorry, this should be `from`. It happens to solve the problem. It looks that the current casting works for the kernel channel is between `n*256 + 1 ~ n*256 + 129`. It fails when it is between `n*256 + 130 ~ n*256 + 256`. For the latter case, we need `sext`.

**Comment by shoubhik** [comment]
@zhiics i think some of the test cases are failing because of this change. We may need to have this conditional if the input is uint.

**Comment by zhiics** [comment]
I tried to dig a little more into this problem by playing with the LLVM IR a bit. I think it might be more of a flag problem because the following test only fail conditionally when the output type is "(u)int32" and output channel c is `8 < c < 256 && c % 2 == 0 && c is not power of 2`. LLVM should try to generate a VNNI instruction when one of the inputs is `int8` and the other is `uint8` and the output is `int32`.

I generated the LLVM IR and experimented it here: https://godbolt.org/z/Cktm2j

It turned out the compilation would fail if you don't specify `-march`, i.e. it could be any of `-march=skylake`, `-march= cascadelake`, `-march=icelake-client, or `-march=icelake-client`. The error message would be the same as what @anijain2305 provided in the discussion forum above.

However, if I choose clang 8.0.0 and `-march=cascadelake` it simply passes, note clang 7.0 will not work because it doesn't support `-march=cascadelake`.  I actually checked the assembly, but I didn't see VNNI instructions there. Instead, I saw many `VPADDD`. I am not sure if it tries to use `mul` and `hadd` to achieve the same functionality (with more instructions). @anijain2305 Can you double check?

```python
data = relay.var("data", shape=(1, 28, 28, 128), dtype='uint8')
kernel = relay.var("w", shape=(3, 3, 128, 256), dtype='int8')
conv = relay.nn.conv2d(data, kernel, kernel_size=(3, 3), out_dtype='int32', data_layout='NHWC', kernel_layout='HWIO')

func = relay.Function([data, kernel], conv)

with relay.build_config(opt_level=0):
    graph, lib, params = relay.build(func, 'llvm', params=None)
ctx = tvm.cpu(0)
module = runtime.create(graph, lib, ctx)
print(module.get_source())
module.run()
```

**Comment by zhiics** [comment]
I will just add a unit test in qnn to help us understand how we can avoid the llvm error.

**Comment by shoubhik** [comment]
Will this work for ARM cpus as well?

**Comment by zhiics** [comment]
@shoubhik This error was from X86 lowering in LLVM. I think ARM may not have the same error though I haven't tried it yet.

File: tests/python/relay/test_op_qnn_conv2d.py [hunk 1]
```diff
@@ -642,6 +642,11 @@ def test_broadcast_layout():
 
 
 def test_conv2d_int8():
+    target = "llvm -mcpu=core-avx2"
+    if not tvm.module.enabled(target):
+        print("skip because %s is not enabled..." % target)
+        return
+
```
Comment (u99127) [inline tests/python/relay/test_op_qnn_conv2d.py]: So, how do we run the same test for Arm or other CPU implementations ? 

I don't think the test infrastructure allows for this kind of execution yet but we probably need to start thinking about this and need helper functions across the test infrastructure that allows us to parameterize this on a per cpu target basis.

Ramana




Ramana

Comment (zhiics) [reply tests/python/relay/test_op_qnn_conv2d.py]: Yes, we currently don't have a good way to check cpu features at runtime. The above is what we do in the unit test. I also think this is not a good way. 

The problem to used solved in this issue should be helpful?
https://github.com/dmlc/tvm/issues/4120

Comment (u99127) [reply tests/python/relay/test_op_qnn_conv2d.py]: Agreed -

 maybe we should create an issue to go clean all of this up once #4120 lands. which was why I wasn't requesting changes :)

Comment (zhiics) [reply tests/python/relay/test_op_qnn_conv2d.py]: Exactly:)

**Comment by icemelon9** [comment]
Thanks @zhiics @anijain2305 @u99127

**Pull request merged**